src/service/rbac-service.ts

```diff
@@ -17,42 +17,246 @@
-                  relation: relationName,
-                  attributes: [...action[relationName]],
-                })),
-              };
-            }),
+            action: actionName,
+            // Map every relation permission to response
+            relations: relationPermissions.map((relationPerm): RelationResponse => ({
+              relation: relationPerm.relation,
+              attributes: relationPerm.attributes,
+            })),
           };
         }),
       };
     });
   }
+
+  /**
+   * Converts the RoleDefinitions object to an Roleresponse, which can be
+   * returned in the API response.
+   * @param roles - The role definitions to parse
+   */
+  public static asRoleResponse(roles: Role[]): RoleResponse[] {
+    return roles.map((role): RoleResponse => {
+      return {
+        id: role.id,
+        name: role.name,
+        systemDefault: role.systemDefault,
+        userTypes: role.userTypes ?? [],
+        // Map every entity permission to response
+        permissions: role.permissions ? this.asEntityResponse(role.permissions) : undefined,
+      };
+    });
+  }
+
+  /**
+   * Convert a human-readable permission definition to a list of
+   * database permission rules
+   * @param def
+   */
+  public static definitionToRules(def: PermissionDefinition): PermissionRule[] {
+    const permissions: PermissionRule[] = [];
+    Object.keys(def).forEach(entity => {
+      Object.keys(def[entity]).forEach((action) => {
+        Object.keys(def[entity][action]).forEach((relation) => {
+          const attributes = Array.from(def[entity][action][relation]);
+          permissions.push({ entity, action, relation, attributes });
+        });
+      });
+    });
+    return permissions;
+  }
+
+  /**
+   * Convert a list of database permission rules to a human-readable
+   * permissions object
+   * @param rules
+   */
+  public static rulesToDefinition(rules: PermissionRule[]): PermissionDefinition {
+    return rules.reduce((permDef: PermissionDefinition, permission1, i1, rolePermissions) => {
+      if (permDef[permission1.entity]) return permDef;
+
+      permDef[permission1.entity] = rolePermissions
+        .filter((p) => p.entity === permission1.entity)
+        .reduce((entDef: EntityDefinition, permission2, i2, entityPermissions) => {
+          if (entDef[permission2.entity]) return entDef;
+
+          entDef[permission2.action] = entityPermissions
+            .filter((p) => p.action === permission2.action)
+            .reduce((actDef: ActionDefinition, permission3) => {
+              if (actDef[permission3.relation]) return;
+
+              actDef[permission3.relation] = new Set(permission3.attributes);
+              return actDef;
+
+            }, {});
+          return entDef;
+
+        }, {});
+      return permDef;
+    }, {});
+  }
+
+  /**
+   * Get a tuple with a list of all roles and the total number of
+   * roles matching the parameters
```

Comment:
Why the count?

Reply:
See the comment above about pagination for this method.